README.txt
= png

== About

png is an almost-pure-ruby PNG library.  It lets you write a PNG without any C
libraries.  It might be a bit "slow", especially if you don't have a C
compiler.

== Installing png

Just install the gem:

  $ sudo gem install png

== Using png

  require 'png'
  
  canvas = PNG::Canvas.new 200, 200
  
  # Set a point to a color
  canvas[100, 100] = PNG::Color::Black
  
  # draw an anti-aliased line
  canvas.line 50, 50, 100, 50, PNG::Color::Blue
  
  png = PNG.new canvas
  png.save 'blah.png'



CHANGES
== 1.1.0

* Fixed bug in PNG::Canvas#each
* Patch from Tom Werner
  * Added PNG::Canvas#inpsect
  * Added PNG::Canvas#to_s
  * Added PNG::Color#to_ascii
* Sped up PNG#to_blob with RubyInline

== 1.0.0

BIRTHDAY!


test/test_png.rb
require 'test/unit'
require 'png'

class TestArray < Test::Unit::TestCase

  def test_fast_pack
    data = [0xff, 0xff, 0xff, 0x00]

    assert_equal data.pack('C*'), data.fast_pack
  end

  def test_fast_flatten
    canvas = [
      [0, [0x01, 0x02, 0x03, 0x04], [0x05, 0x06, 0x07, 0x08]],
      [0, [0x09, 0x0a, 0x0b, 0x0c], [0x0d, 0x0e, 0x0f, 0x10]],
    ]
    assert_equal canvas.flatten, canvas.fast_flatten
  end

end

class TestPNGColor < Test::Unit::TestCase

  def setup
    @color = PNG::Color.new 0x01, 0x02, 0x03, 0x04
  end

  def test_r
    assert_equal 0x01, @color.r
  end

  def test_g
    assert_equal 0x02, @color.g
  end

  def test_b
    assert_equal 0x03, @color.b
  end

  def test_a
    assert_equal 0x04, @color.a
  end

  def test_blend
    c1 = @color
    c2 = PNG::Color.new 0xFF, 0xFE, 0xFD, 0xFC

    assert_equal PNG::Color.new(0xfb, 0xfa, 0xf9, 0xf8), c1.blend(c2)
  end

  def test_intensity
    assert_equal PNG::Color.new(0x01, 0x02, 0x03, 0x3c), @color.intensity(0xf00)
  end

  def test_inspect
    assert_equal "#<PNG::Color 01 02 03 04>", @color.inspect
  end

  def test_to_ascii
    assert_equal '00', PNG::Color::White.to_ascii
    assert_equal '++', PNG::Color::Yellow.to_ascii
    assert_equal ',,', PNG::Color::Red.to_ascii
    assert_equal '..', PNG::Color::Black.to_ascii
  end

end

class TestPNGCanvas < Test::Unit::TestCase

  def setup
    @canvas = PNG::Canvas.new 10, 10
  end

  def test_each
    canvas = PNG::Canvas.new 2, 2
    points = []

    canvas.each do |*data| points << data end

    expected = [
      [0, 0, PNG::Color::White],
      [1, 0, PNG::Color::White],
      [0, 1, PNG::Color::White],
      [1, 1, PNG::Color::White],
    ]

    assert_equal expected, points
  end

  def test_inspect
    assert_equal "#<PNG::Canvas 10x10>", @canvas.inspect
  end

  def test_point
    assert_equal PNG::Color.new(0xfe, 0x00, 0xfe, 0xfe),
                 @canvas.point(5, 5, PNG::Color::Purple)
  end

  def test_line
    @canvas.line 0, 0, 9, 9, PNG::Color::Black

    expected = <<-EOF
..000000000000000000
00..0000000000000000
0000..00000000000000
000000..000000000000
00000000..0000000000
0000000000..00000000
000000000000..000000
00000000000000..0000
0000000000000000..00
000000000000000000..
    EOF

    assert_equal expected, @canvas.to_s
  end

  def test_to_s
    expected = <<-EOF
00000000000000000000
00000000000000000000
00000000000000000000
00000000000000000000
00000000000000000000
00000000000000000000
00000000000000000000
00000000000000000000
00000000000000000000
00000000000000000000
    EOF

    assert_equal expected, @canvas.to_s
  end

end

class TestPNG < Test::Unit::TestCase

  def setup
    @canvas = PNG::Canvas.new 10, 10
    @png = PNG.new @canvas
  end

  def test_class_chunk
    chunk = PNG.chunk 'IDHR', [10, 10, 8, 6, 0, 0, 0 ].pack('N2C5')
    expected = "\000\000\000\rIDHR\000\000\000\n\000\000\000\n\b\006\000\000\000o\345\265\221"
    assert_equal expected, chunk
  end

  def test_to_blob
    expected = <<-EOF.unpack('m*').first
iVBORw0KGgoAAAANSUhEUgAAAAoAAAAKCAYAAACNMs+9AAAAEUlEQVR42mP4
TyRgGFVIX4UAI/uOgGWVNeQAAAAASUVORK5CYII=
    EOF

    assert_equal expected, @png.to_blob
  end

end



lib/png.rb
require 'zlib'

class Array # :nodoc:

  require 'inline'

  inline do |builder|
    builder.include '"intern.h"'

    builder.c <<-EOC
      static void
      fast_flatten() {
        VALUE flat, row, pixel;
        long total_length, height, width, cur = 0, x = 0, y = 0;

        height = RARRAY(self)->len;
        width = RARRAY(RARRAY(self)->ptr[0])->len;
        total_length = height * (width - 1) * 4 + height; /* data + filter */

        flat = rb_ary_new2(total_length);

        for (x = 0; x < height; x++) {
          row = RARRAY(self)->ptr[x];

          pixel = RARRAY(row)->ptr[0]; /* row filter */
          MEMCPY(RARRAY(flat)->ptr + cur, &pixel, VALUE, 1);
          cur++;

          for (y = 1; y < width; y++) { /* row data */
            pixel = RARRAY(row)->ptr[y];
            MEMCPY(RARRAY(flat)->ptr + cur, RARRAY(pixel)->ptr, VALUE, 4);
            cur += 4;
          }
        }
        RARRAY(flat)->len = total_length;
        return flat;
      }
    EOC

    builder.c <<-EOC
      static void
      fast_pack() {
        VALUE res;
        long i;
        char c;

        res = rb_str_buf_new(RARRAY(self)->len);

        for (i = 0; i < RARRAY(self)->len; i++) {
          c = FIX2LONG(RARRAY(self)->ptr[i]);
          rb_str_buf_cat(res, &c, sizeof(char));
        }

        return res;
      }
    EOC
  end

rescue StandardError, LoadError
  
  def fast_pack
    pack 'C*'
  end

  def fast_flatten
    flatten
  end

end

class String # :nodoc:

  ##
  # Calculates a CRC using the algorithm in the PNG specification.

  def png_crc
    unless defined? @@crc then
      @@crc = Array.new(256)
      256.times do |n|
        c = n
        8.times do
          c = (c & 1 == 1) ? 0xedb88320 ^ (c >> 1) : c >> 1
        end
        @@crc[n] = c
      end
    end

    c = 0xffffffff
    each_byte do |b|
      c = @@crc[(c^b) & 0xff] ^ (c >> 8)
    end
    return c ^ 0xffffffff
  end

end

##
# An almost-pure-ruby Portable Network Graphics (PNG) writer.
#
# http://www.libpng.org/pub/png/spec/1.2/
#
# PNG supports:
# + 8 bit truecolor PNGs
#
# PNG does not support:
# + any other color depth
# + extra data chunks
# + filters
#
# = Example
#
#   require 'png'
#   
#   canvas = PNG::Canvas.new 200, 200
#   canvas[100, 100] = PNG::Color::Black
#   canvas.line 50, 50, 100, 50, PNG::Color::Blue
#   png = PNG.new canvas
#   png.save 'blah.png'
#
# = Note
#
# PNG will drop back to a pure-ruby implementation if the
# RubyInline-accelerated methods in Array fail to compile.

class PNG

  ##
  # Creates a PNG chunk of type +type+ that contains +data+.

  def self.chunk(type, data="")
    [data.size, type, data, (type + data).png_crc].pack("Na*a*N")
  end

  ##
  # Creates a new PNG object using +canvas+

  def initialize(canvas)
    @height = canvas.height
    @width = canvas.width
    @bits = 8
    @data = canvas.data
  end

  ##
  # Writes the PNG to +path+.

  def save(path)
    File.open path, 'wb' do |f| f.write to_blob end
  end

  ##
  # Raw PNG data

  def to_blob
    blob = []
    blob << [137, 80, 78, 71, 13, 10, 26, 10].pack("C*") # PNG signature

    blob << PNG.chunk('IHDR',
                      [ @height, @width, @bits, 6, 0, 0, 0 ].pack("N2C5"))
    # 0 == filter type code "none"
    data = @data.map { |row| [0] + row.map { |p| p.values } }.fast_flatten
    blob << PNG.chunk('IDAT', Zlib::Deflate.deflate(data.fast_pack, 9))
    blob << PNG.chunk('IEND', '')
    blob.join
  end

  ##
  # RGBA colors

  class Color

    attr_reader :values

    ##
    # Creates a new color with values +red+, +green+, +blue+, and +alpha+.

    def initialize(red, green, blue, alpha)
      @values = [red, green, blue, alpha]
    end

    ##
    # Transparent white

    Background = Color.new 0xFF, 0xFF, 0xFF, 0x00
 
    White      = Color.new 0xFF, 0xFF, 0xFF, 0xFF
    Black      = Color.new 0x00, 0x00, 0x00, 0xFF
    Gray       = Color.new 0x7F, 0x7F, 0x7F, 0xFF

    Red        = Color.new 0xFF, 0x00, 0x00, 0xFF
    Orange     = Color.new 0xFF, 0xA5, 0x00, 0xFF
    Yellow     = Color.new 0xFF, 0xFF, 0x00, 0xFF
    Green      = Color.new 0x00, 0xFF, 0x00, 0xFF
    Blue       = Color.new 0x00, 0x00, 0xFF, 0xFF
    Purple     = Color.new 0XFF, 0x00, 0xFF, 0xFF

    def ==(other) # :nodoc:
      self.class === other and other.values == values
    end

    ##
    # Red component

    def r; @values[0]; end

    ##
    # Green component

    def g; @values[1]; end

    ##
    # Blue component

    def b; @values[2]; end

    ##
    # Alpha transparency component

    def a; @values[3]; end

    ##
    # Blends +color+ into this color returning a new blended color.

    def blend(color)
      return Color.new((r * (0xFF - color.a) + color.r * color.a) >> 8,
                       (g * (0xFF - color.a) + color.g * color.a) >> 8,
                       (b * (0xFF - color.a) + color.b * color.a) >> 8,
                       (a * (0xFF - color.a) + color.a * color.a) >> 8)
    end

    ##
    # Returns a new color with an alpha value adjusted by +i+.

    def intensity(i)
      return Color.new(r,g,b,(a*i) >> 8)
    end

    def inspect # :nodoc:
      "#<%s %02x %02x %02x %02x>" % [self.class, *@values]
    end

    ##
    # An ASCII representation of this color, almost suitable for making ASCII
    # art!

    def to_ascii
      brightness = (((r + g + b) / 3) * a) / 0xFF
      return '00' if brightness >= 0xc0
      return '++' if brightness >= 0x7F
      return ',,' if brightness >= 0x40
      return '..'
    end

  end

  ##
  # PNG canvas

  class Canvas

    ##
    # Height of the canvas

    attr_reader :height

    ##
    # Width of the canvas

    attr_reader :width

    ##
    # Raw data

    attr_reader :data

    def initialize(height, width, background = Color::White)
      @height = height
      @width = width
      @data = Array.new(@width) { |x| Array.new(@height) { background } }
    end

    ##
    # Retrieves the color of the pixel at (+x+, +y+).

    def [](x, y)
      raise "bad x value #{x} >= #{@height}" if x >= @height
      raise "bad y value #{y} >= #{@width}" if y >= @width
      @data[y][x]
    end

    ##
    # Sets the color of the pixel at (+x+, +y+) to +color+.

    def []=(x, y, color)
      raise "bad x value #{x} >= #{@height}" if x >= @height
      raise "bad y value #{y} >= #{@width}"  if y >= @width
      @data[y][x] = color
    end

    ##
    # Iterates over each pixel in the canvas.

    def each
      @data.each_with_index do |row, y|
        row.each_with_index do |color, x|
          yield x, y, color
        end
      end
    end

    def inspect # :nodoc:
      '#<%s %dx%d>' % [self.class, @width, @height]
    end

    ##
    # Blends +color+ onto the color at point (+x+, +y+).

    def point(x, y, color)
      self[x,y] = self[x,y].blend(color)
    end

    ##
    # Draws a line using Xiaolin Wu's antialiasing technique.
    #
    # http://en.wikipedia.org/wiki/Xiaolin_Wu's_line_algorithm

    def line(x0, y0, x1, y1, color)
      dx = x1 - x0
      sx = dx < 0 ? -1 : 1
      dx *= sx # TODO: abs?
      dy = y1 - y0

      # 'easy' cases
      if dy == 0 then
        Range.new(*[x0,x1].sort).each do |x|
          point(x, y0, color)
        end
        return
      end

      if dx == 0 then
        (y0..y1).each do |y|
          point(x0, y, color)
        end
        return
      end

      if dx == dy then
        Range.new(*[x0,x1].sort).each do |x|
          point(x, y0, color)
          y0 += 1
        end
        return
      end

      # main loop
      point(x0, y0, color)
      e_acc = 0
      if dy > dx then # vertical displacement
        e = (dx << 16) / dy
        (y0...y1-1).each do |i|
          e_acc_temp, e_acc = e_acc, (e_acc + e) & 0xFFFF
          x0 = x0 + sx if (e_acc <= e_acc_temp) 
          w = 0xFF-(e_acc >> 8)
          point(x0, y0, color.intensity(w))
          y0 = y0 + 1
          point(x0 + sx, y0, color.intensity(0xFF-w))
        end
        point(x1, y1, color)
        return
      end

      # horizontal displacement
      e = (dy << 16) / dx
      (x0...(x1-sx)).each do |i|
        e_acc_temp, e_acc = e_acc, (e_acc + e) & 0xFFFF
        y0 += 1 if (e_acc <= e_acc_temp)
        w = 0xFF-(e_acc >> 8)
        point(x0, y0, color.intensity(w))
        x0 += sx
        point(x0, y0 + 1, color.intensity(0xFF-w))
      end
      point(x1, y1, color)
    end

    ##
    # Returns an ASCII representation of this image

    def to_s
      image = []
      scale = (@width / 39) + 1

      @data.each_with_index do |row, x|
        next if x % scale != 0
        row.each_with_index do |color, y|
          next if y % scale != 0
          image << color.to_ascii
        end
        image << "\n"
      end

      return image.join
    end

  end

end

